nixos/modules/virtualisation/railcar.nix

```
@@ -0,0 +1,122 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.railcar;
+  generateUnit = name: containerConfig:
+    let
+      container = pkgs.ociTools.buildContainer {
+        cmd = pkgs.writeShellScript "run.sh" containerConfig.cmd;
+      };
+    in
+      nameValuePair "railcar-${name}" {
+        enable = true;
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+            ExecStart = ''
+              ${cfg.package}/bin/railcar -r ${cfg.stateDir} run ${name} -b ${container}
+            '';
+            Type = containerConfig.runType;
+          };
+      };
+  mount = {
+    options = {
+      type = mkOption {
+        type = types.string;
+        default = "none";
+        description = ''
+        The type of the filesystem to be mounted.
+        Linux: filesystem types supported by the kernel as listed in 
+        `/proc/filesystems` (e.g., "minix", "ext2", "ext3", "jfs", "xfs", 
+        "reiserfs", "msdos", "proc", "nfs", "iso9660"). For bind mounts 
+        (when options include either bind or rbind), the type is a dummy,
+        often "none" (not listed in /proc/filesystems).
+      '';
+      };
+      source = mkOption {
+        type = types.string;
+        description = "Source for the in-container mount";
+      };
+      options = mkOption {
+        type = with types; loaOf (string);
+        default = [ "bind" ];
+        description = ''
+        Mount options of the filesystem to be used.
+        
+        Support optoions are listed in the mount(8) man page. Note that both
+        filesystem-independent and filesystem-specific options are listed.
+      '';
+      };
+    };
+  };
+in
+{
+  options.services.railcar = {
+    enable = mkEnableOption "railcar";
+
+    containers = mkOption {
+      default = {};
+      description = "Declarative container configuration";
+      type = with types; loaOf (submodule ({ name, config, ... }: {
+        options = {
+          cmd = mkOption {
+            type = types.string;
+            description = "Command or script to run inside the container";
+          };
+
+          mounts = mkOption {
+            type = with types; attrsOf (submodule (mount));
```

Think it would be more idiomatic to call `submodule` when you define `mount`, since it’s never going to be used except by being passed to `submodule`.

```suggestion
            type = with types; attrsOf mount;
```